### DISCUSSION TITLE
Various feature and config requests

### DISCUSSION THREAD

**Discussion by ghost** [torrust/torrust-index-gui#605] [Feature requests]
Hey,
Opening a discussion to not clog up issues with requests that might already be implemented.
If worthy of opening an issue for any of the following, let me know.

1. Add middle/right-click functionalities to the site (e.g. opening new tabs, copying links, etc.).
2. Add the option to remove email requirement for account creation.
3. Add the option to view/modify email associated with account (if so chosen).
4. Add the option to remove magnets from the site and tracker entirely.
5. Add the option to remove the light/dark toggle for users, and set a single theme for the entire site.
6. Add the option to modify the 'date' style (e.g. 'Aug. 5th 2024' or  '08/5/2024' instead of 'Monday, August 5, 2024').
7. Add the option to remove sections of torrent pages from user view:
- Currently, torrent pages display boxes for the following: Description, Comment, Creation Date, Created by, Encoding, Files, Trackers, Canonical InfoHash Group.
- It would be useful if options to remove some of them (depending on admin intentions) from view across the entire site.
8. Add config options for site behavior:
- On index default view, add option to disable the behavior which opens a drop-down with the description when clicking on the torrent, rather than opening the torrent page.
- Add option for torrent page boxes (e.g. Description, Creation Date, etc.) to be closed by default instead of opened, requiring users to click to view more.
- Add option to disable/remove/set as default one or the other of the index layouts (e.g. Make 'Default' the default, and remove 'Table' as an option).
- Add option to modify the torrent url settings. Currently, it's set to `torrent/{infohash}/torrent-name`. Option to have it be `torrent/{infohash}`, `torrent/torrent-name`, `torrent/{incremental number}`, `torrent/{custom-name}`, etc. depending on amount and type of torrents hosted on site.
9. Change the heights between lines in the infobox (e.g. infohash and category being closer than size and uploader)
Preview:
![bug (2)](https://github.com/user-attachments/assets/8ab18c13-d8b4-4f08-99a1-f883978824ce)

Thanks,

**Comment by josecelano** [comment]
Hi @ty5e3a45, I will carefully read your comments and reply to each point. We want to release the version 3.0.0 soon (hopefully this month) and start collecting feedback for the roadmap. So it's the perfect timing for your suggestions.

Most of them are very easy to implement and some have already been implemented like "remove email requirement for account creation".

Some of them are similar to other feature requests like https://github.com/torrust/torrust-index-gui/issues/542.

**Comment by josecelano** [comment]
Hi @ty5e3a45, my full reply:

> Hey, Opening a discussion to not clog up issues with requests that might already be implemented. If worthy of opening an issue for any of the following, let me know.
> 
> 1. Add middle/right-click functionalities to the site (e.g. opening new tabs, copying links, etc.).

Can you give a list of the most important ones for you. I guess:

- Open torrent detail in a new tab in "table" or "default" layout (the whole row being a link?). Not sure if possible for the "Default" layout.
- Add a permalink icon in both torrent list items and torrent details.

> 2. Add the option to remove email requirement for account creation.

That option already exists. With this configuration, the email is not required:

```toml
[registration.email]
required = false
verification_required = false
```

If you totally remove the section `[registration.email]`, the email won't even be shown in the registration form.

> 3. Add the option to view/modify email associated with account (if so chosen).

Yes, the user [public profile ](https://github.com/torrust/torrust-index-gui/discussions/90) and settings are essential. I will create an EPIC for that. We only allow users to change their passwords now.

> 4. Add the option to remove magnets from the site and tracker entirely.

Can I ask why? I guess it's fine to have an option to disable them.

> 5. Add the option to remove the light/dark toggle for users, and set a single theme for the entire site.

In general, I prefer this type of choice to be an end-user choice. When we have a user's settings page, that could be a customization option. Or do you want to force it on all users for some reason?

> 6. Add the option to modify the 'date' style (e.g. 'Aug. 5th 2024' or  '08/5/2024' instead of 'Monday, August 5, 2024').

Same as point 5.

> 7. Add the option to remove sections of torrent pages from user view:
> 
> * Currently, torrent pages display boxes for the following: Description, Comment, Creation Date, Created by, Encoding, Files, Trackers, Canonical InfoHash Group.
> * It would be useful if options to remove some of them (depending on admin intentions) from view across the entire site.

I don't see any reason to not include a flag in the configuration for this. Maybe this should be a dynamic configuration, meaning you can change it without restarting the service from the admin panel. I would like to make the TOML config file to big or complex. I would prefer to have feature flags or a configuration panel for the admin.

Anyway, I don't there is too much info yet. But I can't understand why there could be cases where you don't need/use it.

> 
> 8. Add config options for site behavior:
> 
> * On index default view, add option to disable the behavior which opens a drop-down with the description when clicking on the torrent, rather than opening the torrent page.

That was a controversial feature :-). I'm unsure if this should be a user or an admin choice. It's too low-level, maybe, to be a user setting.

> * Add option for torrent page boxes (e.g. Description, Creation Date, etc.) to be closed by default instead of opened, requiring users to click to view more.

I guess your problem is the critical info is at the bottom :-/.

> * Add option to disable/remove/set as default one or the other of the index layouts (e.g. Make 'Default' the default, and remove 'Table' as an option).

It could be a user option. A user asked for quite the opposite.

> * Add option to modify the torrent url settings. Currently, it's set to `torrent/{infohash}/torrent-name`. Option to have it be `torrent/{infohash}`, `torrent/torrent-name`, `torrent/{incremental number}`, `torrent/{custom-name}`, etc. depending on amount and type of torrents hosted on site.
> 

Now, we are using both:

- https://index.torrust-demo.com/torrent/3f4f9cd8e4a44540002e7386151eb759f76dec51/revenge-of-doctor-x
- https://index.torrust-demo.com/torrent/3f4f9cd8e4a44540002e7386151eb759f76dec51

Can you have multiple aliases for the same resource, or would you like to enable/disable some of them? And why does it depend on the amount of torrents?

Adding more aliases should be easier, although it requires rewriting URLs to something like:

- torrent/info-hashes/{infohash}/torrent-name
- torrent/info-hashes/{infohash}
- torrent/titles/torrent-name
- torrent/id/{incremental-number}
- torrent/slug/{custom-name}

We could also build a single URL with all values:

- torrent/{infohash}/{id}/{slug}/{url_encoded_title}

Why do you want this? I would prefer to keep it simple.

> 9. Change the heights between lines in the infobox (e.g. infohash and category being closer than size and uploader)
>    Preview:
>    ![bug (2)](https://private-user-images.githubusercontent.com/155493608/355464557-8ab18c13-d8b4-4f08-99a1-f883978824ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwMjc0OTQsIm5iZiI6MTcyMzAyNzE5NCwicGF0aCI6Ii8xNTU0OTM2MDgvMzU1NDY0NTU3LThhYjE4YzEzLWQ4YjQtNGYwOC05OWExLWY4ODM5Nzg4MjRjZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwN1QxMDM5NTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMzVkYWNmMDI1NzdjNWNlYWM4MmVmZDJjNzI0ZTEzNzdlYzlhYWM2YzNiMzMyYzEwN2VlYmJiMDc2NmEwZTA1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.e6gCmRYwaMVO-oowR-rymPlkTmpNlOCNtzuxMpik42Q)
> 
> Thanks,

Yes, there are a lot of UI improvements to do.

In general, I see these main topics:

1. Admin panel to customize the platform.
  - Add the option to remove magnets from the site and tracker entirely.
  - Add the option to remove the light/dark toggle for users, and set a single theme for the entire site.
  - Add the option to remove sections of torrent pages from user view.
  - Other config options for site behavior:
2. User settings to customize the user experience.
  - Add the option to view/modify email associated with account (if so chosen).
  - Add the option to set the light/dark toggle for users.
  - Add the option to modify the 'date' style (e.g. 'Aug. 5th 2024' or '08/5/2024' instead of 'Monday, August 5, 2024').
3. Improve UX/UI
  - Add middle/right-click functionalities to the site (e.g. opening new tabs, copying links, etc.).
  - Reorganize URLs.

I think you could create discussions for points 1 and 2. Even if they are sometimes related, it will make it easier to follow discussions.

Regarding point 3, I would like a UX/UI expert to review the whole design, including how to add new potential features like more complex/flexible search filters. For minor fixes like the one you mentioned we can just create new issues (bugs).

**Reply by ghost** [reply]
Thank you for your reply @josecelano . The following are some answers to help clarify and answer your comments on my feature requests.
> > 1. Add middle/right-click functionalities to the site (e.g. opening new tabs, copying links, etc.).
> 
> Can you give a list of the most important ones for you. I guess:
> 
>     * Open torrent detail in a new tab in "table" or "default" layout (the whole row being a link?). Not sure if possible for the "Default" layout.
> 
>     * Add a permalink icon in both torrent list items and torrent details.

I'm not sure if we're talking about the same thing. I am referencing browsers recognizing areas on the index as links or clickable items(areas), allowing for browser manipulations such as the following:

<img src="https://github.com/user-attachments/assets/fbd089a2-f040-4c45-8477-b2f118ec9536" width="250" />

Currently when right-clicking, the browser thinks it's a blank surface, thus resulting in:

<img src="https://github.com/user-attachments/assets/f3899055-19ca-4745-b38a-c9fefba1a593" width="250" />

> > 4. Add the option to remove magnets from the site and tracker entirely.
> 
> Can I ask why? I guess it's fine to have an option to disable them.

Magnets work great for public trackers, but in a semi-(or fully)private tracker setting, where passkeys have to be attached to the magnet links, they can be unreliable. In addition, for novice users magnets are easily copied and shared, which isn't something one wants when passkeys are attached to accounts and intended for use by a single user.

> > 5. Add the option to remove the light/dark toggle for users, and set a single theme for the entire site.
> 
> In general, I prefer this type of choice to be an end-user choice. When we have a user's settings page, that could be a customization option. Or do you want to force it on all users for some reason?

Once in the settings, I don't see a reason why it can't exist as a user-chosen preference, but in the mean time I'd rather the UI be as simple as possible (removing the toggle from view) and force set a dark mode, which most users seem to like off the bat.
Once in the config, another admin might decide they'd rather keep it in view on their tracker, or add more than 2 toggles, depending on their use case.

> > 6. Add the option to modify the 'date' style (e.g. 'Aug. 5th 2024' or  '08/5/2024' instead of 'Monday, August 5, 2024').
> 
> Same as point 5.

Similarly to point 5, once in settings, users can set it to their preference, but in the meantime the possibility to force modify the default (e.g. change the long and clutter-some "Weekday, Month, Date, Year" format) would be appreciated.

> > 7. Add the option to remove sections of torrent pages from user view:
> > 
> > 
> > * Currently, torrent pages display boxes for the following: Description, Comment, Creation Date, Created by, Encoding, Files, Trackers, Canonical InfoHash Group.
> > * It would be useful if options to remove some of them (depending on admin intentions) from view across the entire site.
> [..]
> Anyway, I don't there is too much info yet. But I can't understand why there could be cases where you don't need/use it.

For UI simplicity.  While I understand the reasoning behind those boxes, having a config option to remove them without needing to strip it from the code would be useful.
Personally, I'd remove `Comment`, `Creation Date`, `Created by`, `Encoding`, `Trackers`, and `Canonical InfoHash Group` from view, as the hash and date are redundant since already included in the right-hand side panel, and the rest won't matter or be of use to most end-users. Once in config, other admins can modify it to their own use-case.

> > 8. Add config options for site behavior:
> > 
> > 
> > * On index default view, add option to disable the behavior which opens a drop-down with the description when clicking on the torrent, rather than opening the torrent page.
> 
> That was a controversial feature :-). I'm unsure if this should be a user or an admin choice. It's too low-level, maybe, to be a user setting.
> 
> > * Add option for torrent page boxes (e.g. Description, Creation Date, etc.) to be closed by default instead of opened, requiring users to click to view more.
> 
> I guess your problem is the critical info is at the bottom :-/.
> 
> > * Add option to disable/remove/set as default one or the other of the index layouts (e.g. Make 'Default' the default, and remove 'Table' as an option).
> 
> It could be a user option. A user asked for quite the opposite.

Similarly to the previous points, and especially highlighted by your last line, it depends of the use case and admin intentions. For some people, one site behavior will make sense, for others, it won't.
Contextually and as an example, I would want to force users to open torrent pages in order access the download torrent button (by removing magnet/torrent from the index), as the "Description" box will contain information that I won't want them to miss. Because of this, clicking on a torrent and having the description drop-down instead of opening a separate torrent page (which is currently not possible in a different tab, due to issue highlighted in point n.1) hinders that goal. 

> > * Add option to modify the torrent url settings. Currently, it's set to `torrent/{infohash}/torrent-name`. Option to have it be `torrent/{infohash}`, `torrent/torrent-name`, `torrent/{incremental number}`, `torrent/{custom-name}`, etc. depending on amount and type of torrents hosted on site.
> 
> Now, we are using both:
> 
>     * https://index.torrust-demo.com/torrent/3f4f9cd8e4a44540002e7386151eb759f76dec51/revenge-of-doctor-x
> 
>     * https://index.torrust-demo.com/torrent/3f4f9cd8e4a44540002e7386151eb759f76dec51
> 
> 
> Can you have multiple aliases for the same resource, or would you like to enable/disable some of them? And why does it depend on the amount of torrents?
> 
> Adding more aliases should be easier, although it requires rewriting URLs to something like:
> 
>     * torrent/info-hashes/{infohash}/torrent-name
> 
>     * torrent/info-hashes/{infohash}
> 
>     * torrent/titles/torrent-name
> 
>     * torrent/id/{incremental-number}
> 
>     * torrent/slug/{custom-name}
> 
> 
> We could also build a single URL with all values:
> 
>     * torrent/{infohash}/{id}/{slug}/{url_encoded_title}
> 
> 
> Why do you want this? I would prefer to keep it simple.

I hadn't considered aliases, though they seem like a promising idea.

I was thinking about having the option to change the default url format.
Currently, URLs are `*torrent/{hash}/{torrent-name}`.
e.g. https://index.torrust-demo.com/torrent/54cfe9ef983caf4ee8acf7ad66257dd461036f35/alice-in-wonderland-librivox

Having `*torrent/{hash}/{torrent-name}` is long and clutter-some.
With bringing this up, I had in mind the `*torrent/{id}` format where {id} is a number derived from the db (and I believe was the default in previous Torrust versions..?), and keeps it short (e.g. https://index.torrust-demo.com/torrent/5).

However, if you want to go even farther, you can also do the following, but I don't know if it's necessary.
If at their creation, each torrent is assigned (and has stored somewhere) the values for it's {hash}, {id}, {title}, {date} and {torrent-name}, the torrust config could allow for custom URL format (e.g. `*torrent/{id}/{torrent-name}`, or `*torrent/{torrent-name}`) based on admin preferences. 
As an example, [this torrent](https://index.torrust-demo.com/torrent/54cfe9ef983caf4ee8acf7ad66257dd461036f35/alice-in-wonderland-librivox) would have:
`{hash}` = `54cfe9ef983caf4ee8acf7ad66257dd461036f35`
`{title}` = `Alice in wonderland librivox`
`{torrent-name}` = `alice-in-wonderland-librivox`
`{date}` = `2024-04-18`
`{id}` = `5`
and an admin could set-up the default URL format to be anything from:
*torrent/`{hash}`/`{torrent-name}` = */torrent/`54cfe9ef983caf4ee8acf7ad66257dd461036f35`/`alice-in-wonderland-librivox`
*torrent/`{id}`/`{torrent-name}` = *torrent/`5`/`alice-in-wonderland-librivox`
*torrent/`{name}` = *torrent/`alice-in-wonderland-librivox`
and the list goes on.
That way, the default could be set, and those same values could be used for the alias idea you mentioned above where typing *torrent/`%`/ followed by the desired value with `%` being a choice of either `hash`/`id`/`name` would redirect to the default URL format set above, and allow for optional linking methods.

> > 9. Change the heights between lines in the infobox (e.g. infohash and category being closer than size and uploader)
> >    Preview:
> >    ![bug (2)](https://private-user-images.githubusercontent.com/155493608/355464557-8ab18c13-d8b4-4f08-99a1-f883978824ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMwMjc0OTQsIm5iZiI6MTcyMzAyNzE5NCwicGF0aCI6Ii8xNTU0OTM2MDgvMzU1NDY0NTU3LThhYjE4YzEzLWQ4YjQtNGYwOC05OWExLWY4ODM5Nzg4MjRjZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwN1QxMDM5NTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMzVkYWNmMDI1NzdjNWNlYWM4MmVmZDJjNzI0ZTEzNzdlYzlhYWM2YzNiMzMyYzEwN2VlYmJiMDc2NmEwZTA1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.e6gCmRYwaMVO-oowR-rymPlkTmpNlOCNtzuxMpik42Q)
> > 
> > Thanks,
> 
> Yes, there are a lot of UI improvements to do.
> 
> In general, I see these main topics:
> 
>     1. Admin panel to customize the platform.
> 
> 
>     * Add the option to remove magnets from the site and tracker entirely.
> 
>     * Add the option to remove the light/dark toggle for users, and set a single theme for the entire site.
> 
>     * Add the option to remove sections of torrent pages from user view.
> 
>     * Other config options for site behavior:
> 
> 
>     2. User settings to customize the user experience.
> 
> 
>     * Add the option to view/modify email associated with account (if so chosen).
> 
>     * Add the option to set the light/dark toggle for users.
> 
>     * Add the option to modify the 'date' style (e.g. 'Aug. 5th 2024' or '08/5/2024' instead of 'Monday, August 5, 2024').
> 
> 
>     3. Improve UX/UI
> 
> 
>     * Add middle/right-click functionalities to the site (e.g. opening new tabs, copying links, etc.).
> 
>     * Reorganize URLs.
> 
> 
> I think you could create discussions for points 1 and 2. Even if they are sometimes related, it will make it easier to follow discussions.
> 
> Regarding point 3, I would like a UX/UI expert to review the whole design, including how to add new potential features like more complex/flexible search filters. For minor fixes like the one you mentioned we can just create new issues (bugs).

I agree, i'll start discussions for points 1 and 2 within the next few days. For the minor fixes, do you want to create the issues?

**Reply by josecelano** [reply]
Hi @ty5e3a45 thank you for your clarifications! I really appreciate your feedback.

Regarding,

### Links

I agree; I don't like links that only work within JS is enabled. I would like the app to work even when JS is disabled. That's another reason to change them.

### Layout customizations

Regarding the torrent description in the torrent list, there is another reason to disable it: page load statistics. Maybe you want to know what your users are interested in the most; you could count page load for torrent details.

Other customizations are also fine. I'm only afraid that the admin panel to enable/disable them might be too complex.

### URLs

I don't think a URL like this https://index.torrust-demo.com/torrent/54cfe9ef983caf4ee8acf7ad66257dd461036f35/alice-in-wonderland-librivox is too long. We decided to use it for two reasons:

1. We wanted to use a universal identifier (the infohash). If you use the internal database number people can't go directly to a torrent if they know the infohash.
2. We added the "slug" suffix for SEO. And also because it can be more human-friendly than only the infohash.

But I understand you value short URLs. Multiple URls pointing to the same resource are easy to implement but I'm sure about the configurable URL. It looks like a complex feature to implement. I guess it would be easier to  implement if we just have one ID that can have different meanings like:

https://index.torrust-demo.com/torrent/{identifier}

where the identifier can be anything you want using predefined fields:

- `hash`. For example: `54cfe9ef983caf4ee8acf7ad66257dd461036f35`
- `title` For example: `Alice%20in%20wonderland%20librivox` (URL encoded title)
- `slug` For example: `alice-in-wonderland-librivox`
- `date` For example: `2024-04-18`
- `id` For example: `5`

In the configuration, you just need to specify a combination with a separator:

- identifier = hash-slug (https://index.torrust-demo.com/torrent/hash-slug)
- identifier = hash-slug-date
- identifier = hash-slug-id
- etcetera.

cc @torrust/torrustaceans

**Reply by ghost** [reply]
@josecelano Following up on the above. Any further information regarding implementation?

**Reply by josecelano** [reply]
Hi @ty5e3a45, I was focused on the release 3.0.0. Now, we are trying to set the roadmap for next year. See https://github.com/orgs/torrust/discussions/19. I currently have less time available for this project. Unfortunately, I can promise it will be implemented soon. To be honest, I prefer backend tasks, but other people in the team like UX/UI tasks. The thing is, we don't know who is using it or what the key features are that people might be waiting for to start using it. I would like to set a roadmap and, define epics, tasks and try to match users' requested features with our interests. We can even vote for them. My impression is we could have good software for public trackers if we improve a lot of things, but I think we still lack other more critical features.

**Reply by ghost** [reply]
@josecelano, Taking that into account, what type of user feedback are you looking for specifically?
The linked discussion mentions
> [..] suggestions, feature requests, or thoughts on the direction we should take [..]

How detailed do you want it to be? Do you solely want feedback on the future of Torrust as a project, or do you want specific propositions?

**Reply by josecelano** [reply]
I would like to have both high-level and specific propositions from the team and external users, contributors, etc. I want to define the "vision" and break it down into epics, until the leaf level with detailed issue definitions.

In my opinion, we are very far from having a private tracker, at least a fully featured one (forums, chats, peer stats, rewards, etc.). However, can we compete with other alternatives for simple public trackers?

Maybe, it's better to start with the high-level and them continue with the specific ones if you see that your propositions are aligned with the high-level goal. I mean, I don't want to make you spend time defining specific propositions if we focus only on the tracker next year (for example, it's not decided at all). You can do it anyway if you want, but the chances we will have time to implement them are lower.

**Comment by josecelano** [comment]
I'm working on the roadmap and creating issues for these feature requests.

### Subtasks

- [ ] https://github.com/torrust/torrust-index-gui/issues/647
- [ ] https://github.com/torrust/torrust-index-gui/issues/649

**Comment by josecelano** [comment]
Hi @ty5e3a45 I'm creating the issues.

I think you did not explain to me the reason for this:

> Add the option to remove magnets from the site and tracker entirely.

I guess you mean:

- Removing magnet links from the UI
- Remove `magnet_link` field in API responses

**Reply by josecelano** [reply]
And I think your point 9 (_Change the heights between lines in the infobox_) was fixed.

**Reply by josecelano** [reply]
Maybe you want to disable magnet-links because of this: https://github.com/torrust/torrust-index-gui/discussions/85
